notes/plutus-report/costing.tex

```diff
@@ -0,0 +1,339 @@
+Hence they will also be put in the protocol parameters, so that pricing can be changed with only a protocol parameter change.
+
+\paragraph{Choosing the pricing model parameters}
+Choosing the pricing model parameters is a complex problem that may have strategic considerations.
+At minimum, the prices will need to be high enough to avoid DoS attacks (\cref{req:costing-dos}).
+
+\subsubsection{Validating scripts}
+Putting the pieces together, what happens when a script is validated is:
+\begin{itemize}
+\item
+  The ledger uses the pricing model parameters to compute the price for the stated resource budget in Ada, and ensures that the transaction has sufficient fees to cover this, and that it does not cause us to exceed the per-block budget (\cref{req:costing-block-budget}).
+\item
+  The ledger runs the script, passing in the stated budget and the cost model parameters.
+\item
+  The evaluator either terminates with success, or stops early if it runs out of budget.
+\end{itemize}
+
+\subsection{Abstract resource units and resource budgets}
+We use two abstract resource units:
+\begin{itemize}
+\item Abstract time/CPU usage (``Time'')
+\item Abstract peak memory usage (``Space'')
+\end{itemize}
+
+Script execution is limited in both Time and Space.
+The overall limit (budget) is therefore a pair of some amount of Time and Space.
+
+Crucially, the interpretation of the Space limit is that it is a limit on \emph{peak} space usage.
+This is because the limit is mainly to prevent crashes due to exceeding a machine's available RAM, not to compensate the slot leader for RAM usage.
+
+The reason we need to track memory at all is because our evaluators can use unbounded memory.
+This is not true in e.g. Ethereum where there is a (small) stack limit, so programs can be assumed to fit within a constant memory budget.
+Plutus Core has both unbounded integers and unbounded recursion, so it is not hard to use a large amount of memory.
+
+\subsection{Building the cost model}
+The high-level description of the cost model said that we would give costs for ``individual operations''.
+This means that the model is tied to exactly what kind of evaluator we use.
+At the moment we use a fairly standard CEK machine.
+
+\todompj{Ref into PLC section}
+
+Our goal in producing a cost model is to make the computed costs be well-correlated with the real costs.
+Since the pricing model can be used to control how these are paid for, we can stick to trying to track reality.
+Hence we start by benchmarking individual parts of execution, and using statistical methods to infer model parameters from that.
+
+\subsubsection{Mapping between real and abstract units}
+Any benchmarks we do will give results in actual units (e.g. microseconds), not our abstract units.
+So in order to actually use them we need to do one of two things:
+\begin{enumerate}
+\item Set a target mapping for how real units should correspond to abstract units, e.g. 1 Time = 1 microsecond.
```

Comment:
I still think the *basis* for this is an assumption that 1 Time = 1ms. Putting it another way, we could do the entirety of what you said, just sticking with ms. It's all talking about real time. It's then at the end where we decide to wave our wand and declare that the units are actually Time.

Reply:
And yeah, it is all going to be specific to one machine. I should say something about that!

Reply:
> Putting it another way, we could do the entirety of what you said, just sticking with ms.

Not exactly.  We're deciding how to allocate an average time `α` to the compute steps based on dividing the time not accounted for by builtins by the number of compute steps. We have to _count_ the number of compute steps first to do that, and I'm re-using the costing infrastructure to do that by allocating unit cost to each compute step: it's telling us the number of times `compute` is executed, not the time taken to execute them (which we measure independently). Afterwards we'll probably want to set the cost of `compute` to `α` for run-time accounting.  I suppose we're using a step-based cost model to work out the numbers required in the final time-based one.

Also there's some overhead in the builtin benchmarks because they're working out the cost of executing things like `[(builtin multiplyInteger) (con integer 281389012) (con integer 33489021)]` and so they include the time for `con`, `app` and so on.  We may ultimately have to adjust the benchmark costs downwards to account for this: the overall model will give you something like `N ~ βS + αT + σ`, where `β` (which I failed to mention above) takes care of this for you.

Reply:
Sure, we'd be *counting* the steps, but not mentioning our abstract units. The fact that we're using the costing machinery to get the counts doesn't seem like it should matter?

I feel like maybe we're talking past each other: regardless of the details of how we're currently making the benchmarks work, they must be based on some assumption of how the real units and the abstract units match up (the benchmarks alone won't tell us).